### ISSUE TITLE
Transparency improvements

### ISSUE THREAD

**Issue by mtsr** [bevyengine/bevy#2223]
## What problem does this solve or what need does it fill?

Currently transparency is handled by splitting entities into two groups and rendering them in sorted order:
* opaque entities (`Visible::is_transparent == false`)- front-to-back;
* transparent (`Visible::is_transparent == true`) - back-to-front.

Unfortunately this is not enough to handle the general case. One of the simplest problem cases is when rendering a transparent sphere with another transparent inside. Since the sorting happens by entity transform, the object inside will be incorrectly rendered first if it's behind the origin of the sphere. And it will be incorrectly rendered last if it's in front of the origin of the sphere. To render this situation correctly the back of the sphere needs to be rendered first, then the object inside and then the front of the sphere. This cannot be done by sorting entities.

A reasonable introduction to the problem and evaluation of some solutions can be read in [this presentation by nvidia](https://on-demand.gputechconf.com/gtc/2014/presentations/S4385-order-independent-transparency-opengl.pdf). This slide captures the problem in a nutshell:

![image](https://user-images.githubusercontent.com/1319584/118934327-8b41dd00-b94a-11eb-8228-eb42028cc515.png)

## What solution would you like?

I think we need to evaluate the costs/benefits of a few different solutions and maybe even implement more than one of them.

Order Independent Transparency seems the most general solution, but the performance cost can be significant. It seems like Weighted Blended Order Independent Transparency has the best trade-off of the common implementations.

## What alternative(s) have you considered?

There's a number of different possible solutions:
* Depth peeling;
![image](https://user-images.githubusercontent.com/1319584/118934480-b1677d00-b94a-11eb-8020-fd5e3cfdbf54.png)
* Record & sort (in a number of variations, see the nvidia presentation);
![image](https://user-images.githubusercontent.com/1319584/118934799-0c996f80-b94b-11eb-97e1-b9752dd5ec99.png)
* Order Independent Transparency (OIT), by using a commutative blend function, such as [Weighted Blended Order-Independent Transparency (WBOIT) [McGuire et al.]](http://jcgt.org/published/0002/02/09/);

## Additional context

[This demo ](https://github.com/nvpro-samples/vk_order_independent_transparency) shows 7 different algorithms for OIT implemented using Vulkan.

[This video ](https://www.youtube.com/watch?v=ool2E8SQPGU) discusses how FrostBite uses OIT for rendering hair.

**Comment by cart** [comment]
I'm only familiar with this topic in that I know order independent transparency can be expensive and is an active area of research. Unless WBOIT is approximately a drop-in replacement for "traditional transparency" performance-wise, I think it should probably be opt-in / not default. If that is the case, we should pick an algorithm that can be selectively enabled for specific cases that need it.

But yeah, this is definitely something we need to do. Thanks for collecting material for it!

**Comment by CptPotato** [comment]
Weight blended OIT sounded like the most promising while discussing on discord. To me it stands out because it seems to be the easiest to implement and most robust (it's basically a weighted sum over all transparent surfaces - no sorting, no limits to how many transparent layers are possible, etc.). Performance wise I don't see it having much of an impact. The other techniques seem much more elaborate here as they actually perform sorting or some separation of layers. The main downside is that WBOIT is incorrect but it still produces a "plausible" looking result.

Unless I'm missing something, compared to traditional transparency it requires one additional fullscreen rendertarget where everything transparent renders to and one additional compositing pass to resolve the transparency buffer and (post process).

I also thought about it in terms of playing nice with the rest of the rendering and didn't come up with any blockers. Even from the MSAA perspective it should work very well (the compositioning step will also resolve MSAA in this case).

**Comment by mtsr** [comment]
'Traditional' transparency already breaks down for the transparent sphere, though. So it quickly requires extra effort from the user to make it work. Let alone what happens with more complex meshes.

**Comment by cart** [comment]
> 'Traditional' transparency already breaks down for the transparent sphere, though. So it quickly requires extra effort from the user to make it work. Let alone what happens with more complex meshes.

But if order-independent transparency is too expensive, some common cases like "drawing lots of grass on the ground" might be very slow by default (although "traditional transparency" also isn't ideal for that use case afaik). The only thing that I currently know performs well enough to be a default is "traditional transparency". But we should implement as many options as we can, compare them, then make a call. This is a case where giving people lots of options is a good idea anyway, so it wouldn't be wasted effort.

**Comment by aevyrie** [comment]
I think this bug in sponza _might_ be caused by entity transform ordering: 
![image](https://user-images.githubusercontent.com/2632925/119896286-ad5fd000-bef3-11eb-9dc2-104d8330daeb.png)

The column is visible behind the vegetation, but the banner is not. It could be an outright sorting bug though, the banner doesn't have any transparency and should be visible.

This is how it renders in windows 3d viewer:
![image](https://user-images.githubusercontent.com/2632925/119896757-54446c00-bef4-11eb-8feb-953584555fa4.png)

I can open an issue if this is unrelated.

**Comment by alice-i-cecile** [comment]
Closely related to #64.

**Comment by superdump** [comment]
Noting two things:
- The bugs in Sponza are fixed. They used alpha mask materials that are opaque when the alpha is above a threshold and fully transparent below it.
- A noted shortcoming of WBOIT is that it doesn’t work when alpha is close to opaque. Quoting from nvidia:
> However, it can also diverge from the ground truth in scenarios where order strongly affects the result, such as when opacity is high.

**Comment by mtsr** [comment]
This isn't relevant to me right now and it doesn't seem like it's something bevy needs right now either.

**Comment by IceSentry** [comment]
Yep, I'm planning on upstreaming my work during the 0.13 dev cycle

**Comment by alice-i-cecile** [comment]
OIT has been merged; closing as resolved :)
